samples/web_shell/lib/shell/service_worker.js

```diff
@@ -0,0 +1,168 @@
+
+import JsFileSystem from 'axiom/fs/js/file_system';
+import DomFileSystem from 'axiom/fs/dom/file_system';
+import GDriveFileSystem from 'axiom/fs/gdrive/file_system';
+import FileSystemManager from 'axiom/fs/base/file_system_manager';
+import ExecuteContext from 'axiom/fs/base/execute_context';
+import StdioSource from 'axiom/fs/stdio_source';
+import Path from 'axiom/fs/path';
+import AxiomError from 'axiom/core/error';
+
+/**
+ * @constructor
+ *
+ * A Service worker class.
+ *
+ * @param FileSystemManager fsm
+ */
+var ServiceWorker = function(fsm) {
+  this.fsm = fsm;
+};
+
+export {ServiceWorker};
+export default ServiceWorker;
+
+/**
+ *
+ * @return Promise<null>
+ */
+ServiceWorker.prototype.register = function() {
+  if ('serviceWorker' in navigator) {
+    return navigator.serviceWorker.register('/service_worker.js',
+        {scope: '/'}).then(function(reg) {
+      // registration worked
+      console.log('Registration succeeded. Scope is ' + reg.scope);
+    }).catch(function(error) {
+      // registration failed
+      console.log('Registration failed with ' + error);
+    });
+  }
+}
+
+/**
+ * @param {string} message
+ *
+ * @return Promise<@>
+ */
+ServiceWorker.prototype.processMessage = function(message) {
+
+  var url = this.getFileSystemUrl(message);
+
+  return new Promise(function(resolve, reject) {
+    return this.readUrl(url).then(function(data) {
+      var response = this.createResponse(message.subject, data);
+      return this.sendMessage(response);
+    }.bind(this)).catch(function(e) {
+      var response = this.createResponse(message.subject, null, e);
+      return this.sendMessage(response);
+    }.bind(this));
+  }.bind(this));
+};
+
+/**
+ *
+ * @param {string} subject
+ * @param {@} data
+ *
+ * @return {@}
+ */
+ServiceWorker.prototype.createResponse = function(subject, data, error) {
+  var response = {};
+  response.subject = subject;
+  response.name = 'response';
+  response.result = [data];
+  response.error = error;
+  return response;
+};
+
+/**
+ *
+ * @param {string} url
+ *
+ * @return Promise<*>
+ */
+ServiceWorker.prototype.readUrl = function(url) {
+  return new Promise(function(resolve, reject) {
+    return this.fsm.readFile(new Path(url)).then(function(result) {
+      return resolve(result.data);
+    }.bind(this)).catch(function(e) {
+      reject(url + ' : Error in reading file.');
+    });
+  }.bind(this));
+};
+
+/**
+ * @param {@} message
+ *
+ * @return {string}
+ */
+ServiceWorker.prototype.getFileSystemUrl = function(message) {
+  if (!message || message.name !== 'filesystem' || !message.args
+      || !message.args.url) {
+    throw new AxiomError.Invalid(message);
+  }
+
+  var re = /\/local\/.*/;
+  var matches = re.exec(message.args.url);
+  if (!matches || !matches[0].startsWith('/local/')) {
+    throw new AxiomError.Invalid(message);
+  }
+
+  var result = matches[0];
+
+  // Remove /local/ prefix.
+  result = result.substr(7);
+
+  if (result.startsWith('html5')) {
```

Comment:
It would be nicer to use `fsm.getFileSystems()` to know which file systems are available instead of hard-coding "html5" and "gdrive"